### ISSUE TITLE
Improve handling of (code generated) XCode project files

### ISSUE THREAD

**Issue by BPerlakiH** [kiwix/apple#607]
# The origin of the problem

Xcode needs a so called project folder in order to compile an app.
This is a collection of XMLs that is referencing other files, has settings per target and per scheme, and other general compile options.

In order to dynamically work with similar settings to the main repo, but building custom apps, we need to change some of these settings, and might not need to include all the files in the project (eg. selecting only a single language file to be included), or add special files, such as a custom zim file.

Editing the XML file doesn't seem like a viable option here, the most important file is over 1500 lines long.

For these reasons, we wanted to use XCodeGen, which based on a yml project file and the files available in the repository, can create the very same Xcode project directory. It has templating, scheme and other options as well, which is very handy for our goals.

# The current problem

What can cause a change in the project directory? It can be a manual setting change in Xcode, but it can also be a file change, eg. adding a new file / renaming a file or deleting it (outside of the Xcode project folder that is, eg: adding a new Model or Utility class).

We wanted to leave the Xcode project folder in git, so that a new developer can clone the project and start using it. This raises a question how can we make sure that the project file is up to date?

**We wanted to add the project folder to .gitignore. This is not working as we expected.** The project folder is already "tracked" by git, so every change to it needs to be committed. Adding the project folder to .gitignore, has no effect, it is still "tracked" by git. From the [git documentation](https://git-scm.com/docs/gitignore):
> A gitignore file specifies intentionally untracked files that Git should ignore. Files already tracked by Git are not affected;

This means that we are left we 2 options. Either leaving the project file in git, and make sure it's always up to date, or removing it from git.

# 1) Alternative approach keep tracking the project folder, but regenerate on CI side:
In order to make sure that the project folder is up to date, it needs to be re-generated on the CI side, either at the level of building the Pull Request (a successful build means it's a valid project folder) or at the level of the main branch, when merging PRs.
This means we will apply either **automatic commits** on top of pull requests or as part of merging pull requests.

# 2) Alternative approach, remove the project folder from git:
Keeping in mind that we want to make sure that onboarding developers is  easy, we can combine the first steps of getting started with the project into **a single command**, that will both install the necessary tools, automate the project generation and synchronisation for the developers, and downloads CoreKiwix.xcframework at the same time. A single command that needs to be run only once after cloning the project.

**Comment by kelson42** [comment]
@BPerlakiH Thank you for the issue. This is always needed and consequently the problem now is clear: .gitignore as no effect on tracked files.

I can already tell you thatbthe two sokutions you propose are not OK because robot commitimg is to me a super anti-pattern and the second proposal does not respect the requirements.

> We wanted to leave the Xcode project folder in git, so that a new developer can clone the project and start using it. This raises a question how can we make sure that the project file is up to date?

This is the responsability of the developer and the reviewer. We can build helpers if we remarks that this is difficult to secure. Anyway this is a process optimisation which we don't really know if it is necessary. Optimisations - per definition "non madatory" - should be avoided at the early stage of the development of a feature because it slows down everything and let us loose the focus.

> We wanted to add the project folder to .gitignore. This is not working as we expected. The project folder is already "tracked" by git, so every change to it needs to be committed. Adding the project folder to .gitignore, has no effect, it is still "tracked" by git. From the git documentation:
> 
> A gitignore file specifies intentionally untracked files that Git should ignore. Files already tracked by Git are not affected;

We have many other solutions I would like to consider like:
* git hooks
* CI checks/helpers
* https://alphaefficiency.com/git-stop-tracking-file

@rgaudin thoughs?

Let put this problem aside for the moment and move forward with the project.
